### ISSUE TITLE
URxvt + ibus + HiDPI 

### ISSUE THREAD

**Issue by rr-** [ibus/ibus#1890]
1. Set `GDK_SCALE=2`
2. Configure `urxvt` with `Xresources` like this

    ```
    URxvt.inputMethod: ibus
    URxvt.preeditType: OverTheSpot
    ```

3. Open `urxvt`
4. Open `ibus` with <kbd>meta</kbd>+<kbd>space</kbd>
5. Observe how the popup appears at the wrong location

![20161225_013231_dkb](https://cloud.githubusercontent.com/assets/1045476/21469275/1a1b1f34-ca42-11e6-9b9c-779b54c40436.png)

Not sure if this bug within urxvt, or ibus.

Version info:

**ibus**

```
rr-@tornado:~$ ibus version  
IBus 1.5.13
```

**Linux**
```
Linux tornado 4.8.13-1-ARCH #1 SMP PREEMPT Fri Dec 9 07:24:34 CET 2016 x86_64 GNU/Linux
```

**urxvt**

```
rr-@tornado:~$ urxvt -h 
rxvt-unicode (urxvt) v9.22 - released: 2015-10-11
options: perl,xft,styles,combining,blink,iso14755,encodings=eu+vn+jp+jp-ext+kr+zh+zh-ext,fade,transparent,tint,pixbuf,XIM,frills,selectionscrolling,wheel,slipwheel,cursorBlink,pointerBlank,scrollbars=plain+rxvt+NeXT+xterm
Usage: urxvt [-help] [--help]
 [-display string] [-tn string] [-geometry geometry] [-C] [-iconic]
 [-cd string] [-dockapp] [-/+rv] [-/+ls] [-mc number] [-/+j] [-/+ss] [-/+ptab]
 [-/+sb] [-/+sr] [-/+st] [-sbt number] [-/+si] [-/+sk] [-/+sw] [-/+ip] [-/+tr]
 [-tint color] [-sh number] [-blr HxV] [-fade number] [-fadecolor color]
 [-/+ut] [-/+vb] [-/+tcw] [-/+insecure] [-/+uc] [-/+bc] [-/+pb] [-bg color]
 [-fg color] [-hc color] [-cr color] [-pr color] [-pr2 color] [-bd color]
 [-pixmap file[;geom]] [-icon file] [-fn fontname] [-fb fontname]
 [-fi fontname] [-fbi fontname] [-/+is] [-im name] [-pt style]
 [-imlocale string] [-imfont fontname] [-name string] [-title string]
 [-n string] [-sl number] [-embed windowid] [-depth number] [-visual number]
 [-/+override-redirect] [-pty-fd fileno] [-/+hold] [-w number] [-b number]
 [-/+bl] [-lsp number] [-letsp number] [-/+sbg] [-mod modifier] [-/+ssc]
 [-/+ssr] [-pe string] [-e command arg ...]
```

**Comment by fujiwarat** [comment]
I don't see any differences with ``GDK_SCALE=2`` and don't think urxvt is a GDK application.

**Comment by rr-** [comment]
Maybe it could work like this?

1. `GDK_SCALE=2` scales things 2x while maintaining pixel coordinates
2. `urxvt` requests IBus-Anthy to render at coordinate X,Y
3. IBus-Anthy receives X,Y and draws itself at these coordinates
4. But because of the scaling, while X,Y logically remain the same, physically they get rendered at 2X,2Y

**Comment by rr-** [comment]
With an advice from URxvt folks I tried also xterm (why didn't I think about this before) and the same problem happens there:

![20170101_185844_bom](https://cloud.githubusercontent.com/assets/1045476/21582364/666f0d2e-d054-11e6-96f5-4ec85b336897.png)

**Comment by rr-** [comment]
Following patch to `ibus` seems to help on my particular machine:

```diff
diff --git a/client/x11/main.c b/client/x11/main.c
index a717a2c8..c1bb5673 100644
--- a/client/x11/main.c
+++ b/client/x11/main.c
@@ -293,8 +293,8 @@ _xim_store_ic_values (X11IC *x11ic, IMChangeICStruct *call_data)
     for (i = 0; i < (int)call_data->preedit_attr_num; ++i, ++pre_attr) {
         if (g_strcmp0 (XNSpotLocation, pre_attr->name) == 0) {
             x11ic->has_preedit_area = TRUE;
-            x11ic->preedit_area.x = ((XPoint *)pre_attr->value)->x;
-            x11ic->preedit_area.y = ((XPoint *)pre_attr->value)->y;
+            x11ic->preedit_area.x = ((XPoint *)pre_attr->value)->x/2;
+            x11ic->preedit_area.y = ((XPoint *)pre_attr->value)->y/2;
         }
         else {
             LOG (1, "Unknown preedit attribute: %s", pre_attr->name);
```

perhaps `ibus` could read `GDK_SCALE` and use its value here?

**Comment by fujiwarat** [comment]
I think your problem is not changed either with GDK_SCALE=1, GDK_SCALE=2 or GDK_SCALE=3?

**Comment by fujiwarat** [comment]
OK, I understood your problem.
You set GDK_SCALE for ibus-daemon but not the client applications.
I think your result is expected because the ibus lookup window is made from GTK.
I expect if you enlarge GTK applications generally, you also will enlarge the X11 applications.
Hence I'd like to close this issue as not a bug.

**Comment by rr-** [comment]
Hmm. I am setting the `GDK_SCALE` in `.xinitrc`, so it's available for everyone including the window manager.

```
export GDK_SCALE=2
export QT_SCALE_FACTOR=2

exec bspwm  # window manager

dconf load /desktop/ibus/<~/.config/ibus/config.ini
killall ibus-daemon
ibus-daemon -drR --xim
export GTK_IM_MODULE=ibus
export XMODIFIERS=@im=ibus
export QT_IM_MODULE=ibus
```

**Comment by fujiwarat** [comment]
So you should choose, enlarge X11 applications too, disable to enlarge GDK, set GDK_SCALE=1 for ibus-daemon.

**Comment by rr-** [comment]
The DPI is already set to 140 though. Without additional `GDK_SCALE=2`, Firefox and ibus are super-tiny.

**Comment by fujiwarat** [comment]
If DPI is changed but GTK3 size is small, I think it's a GTK3 issue but not IBus.
I think Firefox is not a good example since it has been GTK2. Probably it's good to try gedit for you.
I think IBus does not have any special implementations in the GTK3 scaling.

ibus-setup has a configuration of the font size.
If you use GTK 3.22 or later, the fix of issue #1879 is required.
